### PR TITLE
Reimplement `static-timing` pass.

### PR DESCRIPTION
**Pull request by mikeurbach**
This is intended to address #654.

### CODE REVIEW

File: calyx/src/passes/top_down_static_timing.rs
Comment (mikeurbach) [inline calyx/src/passes/top_down_static_timing.rs:206]: This got a little weird, and I haven't come back to see if there is a smarter way to address what I saw. The current algorithm seems to work well, but it wouldn't handle the final statement in a seq, so I added this here to address that. But this ended up making duplicate transitions in some cases (really, the only transition this line is needed to add is the final statement in the outermost seq, from what I saw). Perhaps there is a better way to address that case; what I chose was to make the transitions a set, rather than a vector, so any duplicates would be ignored.

File: calyx/src/passes/top_down_static_timing.rs
Comment (mikeurbach) [inline calyx/src/passes/top_down_static_timing.rs:349]: I wasn't sure, do we want to assert all nodes have a static attribute, or does it only matter for enables? I haven't looked into how that attribute is inferred, but from what I've tested, if all the enables are static, the whole control program becomes static.

**Review by mikeurbach — COMMENTED** [review]
This should be in a comment, but while it is still cooling, I am leaving that off. The current approach is roughly this:

* Enables go from start state to start state + time, and have transitions to cur_state + time - 1.
* Seqs create transitions from the start state to the first child, and from each child to the next.
* Pars create transitions from the start state up until the last state of any child.
* Ifs do not create transitions, but return the end state of both branches.
* Whiles are WIP.

Please let me know if there is anything else to do about the correctness static timing tests... they appear to target the same tests as correctness dynamic. I will take a look at enabling the existing passes/static-timing tests, and adding a few more from the issue and the program I'm compiling as an example.

Finally, please apologize my Rust. I've read plenty of Rust before, but this is the first time I've had to write anything significant. I know enough about the borrow checker to know I can spray and pray `.clone()` when I don't understand what it's telling me... I tried to learn some patterns from tdcc, but I expect there are plenty of Rust noob mistakes, so I apologize in advance for that.

File: calyx/src/passes/top_down_static_timing.rs
Comment (mikeurbach) [reply calyx/src/passes/top_down_static_timing.rs:334]: That makes sense, and I was planning to assert this everywhere to expect `infer-static-timing` had been run. I'm curious if that can be relaxed though... unless I'm missing something, the current implementation only really needs to have the static attribute on enables. Less requirements seems like a good thing if that is sufficient.

Comment (mikeurbach) [reply calyx/src/passes/top_down_static_timing.rs:334]: Now that I'm starting to think about how this slots in for tests, I'm curious what the best approach is. Should this pass just ignore IR it can't handle, and leave it to tdcc to deal with later? Or should we defensively error out and expect this pass to not be enabled in situations it can't work?

**Comment by mikeurbach** [comment]
I took a crack at adding the passes/static-timing tests. Now I'm not sure what was meant by the issue about "Enable "correctness static" tests and Add -d static-timing flag to "correctness dynamic" tests". As far as I can see from what I uncommented, those are the same sets of tests? Which ones in the correctness directory should have static timing disabled?

File: calyx/src/passes/top_down_static_timing.rs
Comment (mikeurbach) [reply calyx/src/passes/top_down_static_timing.rs:338]: A couple more points here: apparently while constructs don't receive a @static annotation even when all children do. In turn, this causes a seq generated by remove-comb-groups to not receive a @static annotation. So I think it makes sense to only enforce this on enables. The question remains, what if all enables are not @static, should it error out or do nothing?

Comment (rachitnigam) [reply calyx/src/passes/top_down_static_timing.rs]: While nodes receive a static annotation when they have a @bound annotation on them and all their children have @static annotation too. Without bound we don’t know how many times the loop executes.

Comment (rachitnigam) [reply calyx/src/passes/top_down_static_timing.rs]: This pass is considered an optimization and should in general bail out when it can’t do anything. However, for tests, it makes sense to provide a flag that throws an error instead of silently passing.

Comment (rachitnigam) [reply calyx/src/passes/top_down_static_timing.rs]: Also yes you are correct about only enables needing static annotations. Infer static timing propagates them to parents do that you don’t have to traverse the tree to decide if you can compile a control sub program. Either option is okay with me so up to you.

Comment (mikeurbach) [reply calyx/src/passes/top_down_static_timing.rs]: If this is just an optimization then I will make it do nothing if the whole progam isn't @static. I'm not sure yet where to get the @bound attribute. Is there anything to infer that, or should it be supplied from the frontend?

Comment (rachitnigam) [reply calyx/src/passes/top_down_static_timing.rs]: Generally frontends are expected to provide it. We don’t have a pass to infer it but could be cool…

Comment (mikeurbach) [reply calyx/src/passes/top_down_static_timing.rs]: Ok, I will make sure we stamp it out from CIRCT; it's known statically in every program we support pipelining.

**Comment by mikeurbach** [comment]
@rachitnigam I think this is good for a first draft whenever you get a chance to take a look. I've tested this with the pipelined dot product I get out of CIRCT and it does work and show the stages executing concurrently in a VCD trace!

Screenshot from 2022-02-11 00-42-14

You can see the multiplier, which would be the bottleneck, still isn't fully utilized at steady state... I'm trying to see if I can tighten up the while loop state machine. There's also the register update for the while condition that will always take an extra cycle in the steady state. I haven't thought about the implications, but we could potentially read the condition combinationally to save that.

BTW, I didn't end up using the bounds attribute in the implementation, so the only requirement here is the enables are all static.

**Comment by rachitnigam** [comment]
> I think this is good for a first draft whenever you get a chance to take a look. 

All right, marking this as "ready for review" and asking other folks to join in! Thanks for the work!

> I'm trying to see if I can tighten up the while loop state machine

I think you may have to use the `@bound` attribute for this because `remove-comb-groups` transforms the control program of a `while` operator to execute a group that evaluates the condition before starting the next iteration which probably causes the delay you're seeing. In general, I think `remove-comb-group` should be made into an optional pass and compilation passes should directly deal with combinational groups for conditional evaluation.

**Review by rachitnigam — APPROVED** [review]
Github actions are not running for some reason. Marking the PR as "approved" to see if that helps. Please wait for more comments separately.

**Comment by mikeurbach** [comment]
> In general, I think remove-comb-group should be made into an optional pass and compilation passes should directly deal with combinational groups for conditional evaluation.

Ok great, that is what I was pondering last night. I think we just need to make sure the assertions in the handling for if and while here take that into account; they are currently enforcing remove-comb-groups was run (since I copied them from tdcc). I'll worry about that as a later optimization... for now I am just stoked the "pipelined" approach is indeed "pipelining".

I was about to share this. Here's what the "non-pipelined" version looks like. This is the same program, also going through CIRCT, but just through the SCF dialect, without scheduling at the Affine level. You can see the next multiplies don't begin until after the previous adds finish, whereas in the above the next multiply starts on the same cycle as the previous add.

Screenshot from 2022-02-11 09-42-55

By my reading in gtkwave, the pipelined version has the result ready on the output of the adder after 61 clocks, compared to 93 clocks for the non-pipelined version. I'll try to run a synthesis for some device and get a resource report as well. I'm quite curious how area efficient the top-down-st FSM implementation is.

**Comment by rachitnigam** [comment]
@mikeurbach looks like Github Actions are being annoying about running tests on forks. Can we try one of two things:
1. Close and reopen this PR?
2. If that doesn't help, I'll make you a contributor to this repository and you can open a PR from a branch within this repository?

**Comment by mikeurbach** [comment]
I will try repopening a new on.

**Comment by rachitnigam** [comment]
Okay, I've sent you an invite to collaborate on this repository.

**Pull request closed without merging**
